### ISSUE TITLE
Batch normalization for HyperNEAT networks

### ISSUE THREAD

**Issue by schrum2** [schrum2/MM-NEAT#162]
Adding batch normalization ( https://towardsdatascience.com/batch-normalization-in-neural-networks-1ac91516821c ) to HyperNEAT's substrate networks could stop them from having saturated activations as we saw recently with the newly completed deep architecture you were working with.

However, implementing this with TWEANNs could be a hassle, because they aren't really organized into layers. This is why getting the #112 test to work is important.

**Comment by DevonFulcher** [comment]
Could you point me in the right direction for this one?

**Comment by schrum2** [comment]
Put this one on hold. I need to give it more thought myself.

**Comment by schrum2** [comment]
I've been looking carefully at this, and I think that I initially misunderstood how this worked. However, given my new understanding, I'm having difficulty seeing how we can implement this, since we are not training the network in batches.

The normalization is supposed to happen on each activation independently. Perhaps a better way to say this is that it happens with respect to each individual neuron. Normally, in supervised learning, networks are trained in batches. This means that the network (and therefore all of its neurons) process several inputs before any learning occurs. So, if you present 20 training examples in a batch, each neuron of the network would propagate 20 values. 

Batch normalization means averaging each neuron's outputs across those 20 values during training, and once the network is fully trained the activations are normalized across the statistics of the entire data set. I'm not sure we can easily do this. We don't train in batches.

I do have an idea for something we could to that would be similar to batch normalization, but it is pretty complicated, so I'll describe it in a separate comment later.

**Comment by schrum2** [comment]
This idea will require changing TWEANN.Node and extending TWEANNGenotype.FullNodeGene. This touches some important code, so you can't really work on anything else while focusing on this. 

The idea is to allow each neuron (TWEANN.Node) to have an array that remembers the last N activations is wanted to output (N would be a new command line parameter: batchNormMemoryLength). However, when the network is actually activated, the value that the neuron passes forward would be the normalized activation with respect to all values in the memory. The mean and standard deviation of the activations in the memory can be computed as needed, but there are also two additional parameters needed in each node: In the site references above they are called gamma and beta. These would need to be additional parameters in each node that would be learned by evolution. These parameters would need to be in TWEANN.Node, but because they are learned by evolution, they would also need to be in TWEANNGenotype.

Specifically, a new BatchNormNodeGene should extend FullNodeGene in TWEANNGenotype and add these parameters. We would also need to add mutations that change these values.

It's a pretty ambitious final issue to tackle. It might be better to knock out some of the easier remaining issues with the "complete before end of SCOPE tag" first. Still, I am willing to help if you have questions.

**Comment by DevonFulcher** [comment]
Sounds great! I think we should call the parameter something like nodeNormMemoryLength since it is no longer about batches. The original idea was for this to only be a feature of HyperNEAT but since we are changing the node, this could be used in regular NEAT, right?

**Comment by DevonFulcher** [comment]
Should the nodeNormalizedMemory be an array or a queue?

**Comment by schrum2** [comment]
It could be used in regular NEAT too, but we need to be careful. For example, when using it with HyperNEAT, we only want the substrate TWEANN nodes to have memory, not the CPPN nodes. Let's focus on applying this to HyperNEAT first, but it should be easy for it to work with regular NEAT as well.

I'm actually wondering if we want to extend TWEANN.Node rather than add private members to it. Will require some thought, though we can just try it one way and refactor later.

**Comment by schrum2** [comment]
A bit of both. Normally, a proper queue implementation disallows access to all data except the first and last values. However, we still need to access arbitrary elements to compute an average. So, it will be an array, but we will sort of access it as a queue, at least sometimes. 

After looking at the options a bit, the class I would like you to use is the ArrayDeque(). It has a size() method for checking if we have reached capacity. By the way, be sure to initialize the ArrayDeque to have a capacity equal to the max memory length so that we never incur a resize operation. The ArrayDeque also has an iterator that can be used to compute the average over the elements that are present. This also means that if the memory is not full yet, we can still easily compute an average without having to check how many elements are actually present.

**Comment by schrum2** [comment]
For incremental calculations:
http://datagenetics.com/blog/november22017/index.html

This will allow us to avoid storing a memory entirely and just compute averages across the entire history of inputs

**Comment by DevonFulcher** [comment]
Now that vanilla normalized nodes is working, we need to turn of normalization for inputs and evolve values for gamma and beta.

**Comment by DevonFulcher** [comment]
Post watched normalized nodes with evolved gamma and beta and everything seems normal. Although with gamma and beta being evolved the activation has the potential of going outside the bounds of [-1,1]. I believe you said that this was okay but I just wanted to make sure.

**Comment by DevonFulcher** [comment]
Post watched weekend tests and everything seems fine. Normalized nodes did not perform well and they performed even worse with evolved gamma and beta.
